### DISCUSSION TITLE
Some CSP problems

### DISCUSSION THREAD

**Discussion by bycixman35** [Baroshem/nuxt-security#569] [Q&A]
Hello everyone, there are 2 problems with my code below. No matter what I did, I couldn't solve the problem.

Problem 1 -) When the user refreshes the page, it always redirects to the home page.

Problem 2 -) The css does not come correctly on the pages where I pull dynamic data. I added 2 images as an example. When I refresh the page the css becomes even more corrupted.

Thanks for your help.

![hatalı](https://github.com/user-attachments/assets/ecf494a1-fd79-4436-bc83-d808839cbfde)

![normal](https://github.com/user-attachments/assets/a0bceebd-9d21-4dbb-b18f-7042a29f327b)

`  security: {
    hidePoweredBy: true,
    corsHandler: {
      origin: "https://www.abc.com",
      methods: ["GET", "POST", "DELETE"],
      credentials: false
    },
    headers: {
      strictTransportSecurity: {
        maxAge: 15552000,
        includeSubdomains: true
      },
      xFrameOptions: "DENY",
      crossOriginEmbedderPolicy: "unsafe-none",
      crossOriginOpenerPolicy: "same-origin",
      crossOriginResourcePolicy: "same-origin",
      xXSSProtection: "1; mod=block",
      xPermittedCrossDomainPolicies: "none",
      contentSecurityPolicy: {
        'base-uri': ["'none'"],
        'font-src': ["'self'", 'https:', 'data:'],
        'form-action': ["'self'"],
        'img-src': ["'self'", 'https://www.facebook.com', 'https://www.google.com.tr', 'https://www.googletagmanager.com', 'https://i.ytimg.com', 'data:'],
        'frame-ancestors': ["'none'"],
        'frame-src': ["'self'", 'https://www.youtube.com/', 'https://td.doubleclick.net/'],
        'manifest-src': ["'self'"],
        'media-src': ["'self'"],
        'object-src': ["'none'"],
        "worker-src": ["'self'"],
        "script-src-attr": ["'none'"],
        'script-src': [
          "'self'",
          'https:',
          "'unsafe-inline'",
          "'nonce-{{nonce}}'"
        ],
        'style-src': ["'self'", "'unsafe-inline'"],
        "upgrade-insecure-requests": true
      },
      originAgentCluster: "?1",
      xContentTypeOptions: "nosniff",
      referrerPolicy: "strict-origin-when-cross-origin",
      xDownloadOptions: "noopen",
      xDNSPrefetchControl: "off"
    },
    allowedMethodsRestricter: {
      methods: ["GET", "POST", "DELETE"]
    },
    xssValidator: {
      methods: ["GET", "POST", "DELETE"],
      css: true,
      escapeHtml: true
    },
    requestSizeLimiter: {
      maxRequestSizeInBytes: 2000000,
      maxUploadFileRequestInBytes: 8000000
    },
    rateLimiter: {
      tokensPerInterval: 100,
      interval: 900000,
      headers: false
    },
    sri: true,
    nonce: true,
    ssg: {
      meta: true,
      hashScripts: true,
      hashStyles: false,
      exportToPresets: true
    }
  }`

**Comment by vejja** [comment]
Hi @bycixman35 
Can you please give us a screenshot of your console outputs in the chrome developer tools?

**Comment by bycixman35** [comment]
There is no error in the console.

When I remove nitro prerender routes / the problem is solved, but when I add them, they do not work together.

**Comment by vejja** [comment]
Strange that you don’t have messages in the console
How do you know it’s a CSP issue?

**Reply by bycixman35** [reply]
As I said, when I do nitro prerender routes / it gives the errors I mentioned.

**Comment by vejja** [comment]
This is probably related to #576 
Investigation ongoing

**Comment by vejja** [comment]
@bycixman35 we are releasing a hotfix with PR #577 
This might (or might not) be related to your issue. Would you be able to test ?
Also, can you give me your nuxt.config.ts please so that I can verify what you have in common with #576

**Reply by bycixman35** [reply]
nitro: { compressPublicAssets: true, minify: true, publicAssets: [{ dir: "public/img", maxAge: 86400 }], prerender: { routes: ['/']}},

**Reply by bycixman35** [reply]
I will write to you after testing it.

**Reply by vejja** [reply]
Hold on
I'm releasing v2.1.5
You can test by upgrading to v2.1.5

**Reply by vejja** [reply]
Issue in #576 was the use of minify
I can see you are using minify
Can you try removing this option ?
Also, I would need your full nuxt.config.ts to understand your setup

**Reply by bycixman35** [reply]
I test right now. If I send you the link to the site, can you look at it? There is no virus, don't worry. I want you to see the error.

**Reply by bycixman35** [reply]
When I test it in incognito mode, there is no problem, but it still takes me to the home page when I refresh the page. Since I am using Nuxt3 pwa, I need to do a prerender route /. If I don't, it gives an error.

The CSS problem seems to be solved.

**Reply by bycixman35** [reply]
`  security: {
    hidePoweredBy: true,
    corsHandler: {
      origin: "https://www.abc.com",
      methods: ["GET", "POST", "DELETE"],
      credentials: false
    },
    headers: {
      strictTransportSecurity: {
        maxAge: 15552000,
        includeSubdomains: true
      },
      xFrameOptions: "DENY",
      crossOriginEmbedderPolicy: "unsafe-none",
      crossOriginOpenerPolicy: "same-origin",
      crossOriginResourcePolicy: "same-origin",
      xXSSProtection: "0",
      xPermittedCrossDomainPolicies: "none",
      contentSecurityPolicy: {
        'base-uri': ["'none'"],
        'font-src': ["'self'", 'https:', 'data:'],
        'form-action': ["'self'"],
        'img-src': ["'self'", 'data:'],
        'frame-ancestors': ["'none'"],
        'frame-src': ["'self'"],
        'manifest-src': ["'self'"],
        'media-src': ["'self'"],
        'object-src': ["'none'"],
        "worker-src": ["'self'"],
        "script-src-attr": ["'none'"],
        'script-src': [
          "'self'",
          'https:',
          "'unsafe-inline'",
          "'nonce-{{nonce}}'"
        ],
        'style-src': ["'self'", "'unsafe-inline'"],
        "upgrade-insecure-requests": true
      },
      originAgentCluster: "?1",
      xContentTypeOptions: "nosniff",
      referrerPolicy: "strict-origin-when-cross-origin",
      xDownloadOptions: "noopen",
      xDNSPrefetchControl: "off"
    },
    allowedMethodsRestricter: {
      methods: ["GET", "POST", "DELETE"]
    },
    xssValidator: {
      methods: ["GET", "POST", "DELETE"],
      css: true,
      escapeHtml: true
    },
    requestSizeLimiter: {
      maxRequestSizeInBytes: 2000000,
      maxUploadFileRequestInBytes: 8000000
    },
    rateLimiter: {
      tokensPerInterval: 100,
      interval: 900000,
      headers: false
    },
    sri: true,
    nonce: true,
    ssg: {
      meta: true,
      hashScripts: true,
      hashStyles: false,
      exportToPresets: true
    }
  }`

**Reply by vejja** [reply]
Yes you can send your link thanks 🙏🏻
Can you please copy the output of `npx nuxi info` here ?

**Reply by bycixman35** [reply]
- Operating System: Windows_NT
- Node Version:     v23.2.0
- Nuxt Version:     3.14.1592
- CLI Version:      3.16.0
- Nitro Version:    2.10.4
- Package Manager:  npm@10.9.0
- Builder:          -
- User Config:      default
- Runtime Modules:  @vite-pwa/nuxt@0.10.6, nuxt-purgecss@2.0.0, nuxt-security@2.1.5, @nuxtjs/i18n@9.1.0
- Build Modules:    -

**Reply by bycixman35** [reply]
![image](https://github.com/user-attachments/assets/dcf1af76-302f-416e-9ca6-e1b4b001c220)

If you go in and check the site right now, there is an error in the image on the site.

I solve the error you see in the image by doing prerender routes / but when I integrate nuxt3 security, it redirects to the home page every time the site refreshes.

When I remove nuxt3 security, the problem of redirecting to the home page is gone.

I think the problem is that nuxt3 security and nitro prerender route do not work correctly together.

**Reply by vejja** [reply]
You have two different issues.
The first one is:
<img width="596" alt="Screenshot 2024-11-30 at 17 09 59" src="https://github.com/user-attachments/assets/b0b3ceb9-0a7b-407e-a22b-0e365ae36dfa">
This is a PWA error. Look at the documentation of vite-pwa/nuxt to solve.

The second one is: 
<img width="595" alt="Screenshot 2024-11-30 at 17 24 32" src="https://github.com/user-attachments/assets/cf71ad3c-64df-475f-84da-f1f98bb90753">
This is due to a wrong CSP setting. To solve, start with default values for the `security` option : in order to do this, do not set the `security` object at all. Then find out what are the remaining CSP issues in the console.

Update:
The redirect issue is likely related to the PWA not caching the root url.
The CSP issue is only seen by me, because I'm accessing Google from France, and you only whitelisted the Turkish version of Google. I would advise to use the default settings for `security`, unless you have very specific requirements and you know what you are doing 🥲

**Reply by bycixman35** [reply]
Why does the prerender route work fine when I uninstall nuxt3 security?

**Reply by vejja** [reply]
Not sure. I would need a reproduction to tell...

**Reply by bycixman35** [reply]
How did you solve the CSS problem? What was the problem? I'm asking out of curiosity.

**Reply by vejja** [reply]
The issue was linked to our `removeLoggers` setting.
When set to `true` (which is the default), we were switching on the `vite.build.minify` option.
OP was using Nuxt UI v3-alpha, and for some reason the minification process erases a Tailwind layer in the production build of Nuxt UI.
